### ISSUE TITLE
Please don't use 'Thread.Sleep(TimeSpan.FromSeconds(seconds));' in the examples.

### ISSUE THREAD

**Issue by sablanc-msft** [MicrosoftDocs/powerapps-docs#2633]
[Enter feedback here]
Please don't use 'Thread.Sleep(TimeSpan.FromSeconds(seconds));' in the examples. This can lead to serious issues if used in production. Use await Task.Delay() instead.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 0a6c11a1-c9f6-a904-f49a-0fedbe9f511f
* Version Independent ID: 7db20846-7922-23d4-f261-4c54e61b6913
* Content: [Service protection API limits (Microsoft Dataverse) - Power Apps](https://docs.microsoft.com/en-us/powerapps/developer/data-platform/api-limits#retry-operations)
* Content Source: [powerapps-docs/developer/data-platform/api-limits.md](https://github.com/MicrosoftDocs/powerapps-docs/blob/main/powerapps-docs/developer/data-platform/api-limits.md)
* Service: **powerapps**
* Sub-service: **dataverse-developer**
* GitHub Login: @JimDaly
* Microsoft Alias: **jdaly**

**Comment by phecke** [comment]
@sablanc-msft  We are not expecting customers to use our sample code as-is in production. All sample code is use at your own risk. The samples are written to teach use of the API and do not represent the best way to code.

I will keep your suggestion in mind when we refresh the samples. Our older samples are not written to be asynchronous so they will take more work. However, the newer code CDSWebApiService can be easily changed. I will make the change there for now.

Thanks for the suggestion.

**Comment by JimDaly** [comment]
@sablanc-msft 
Can you provide a bit more justification for the kinds of serious issues that can occur in production?
We do need to expect that customers may use our sample code as-is in production and it should represent the best practices.
Using Thread.Sleep was reviewed by several devs and didn't cause them to raise an issue.

I will consider applying this change if I understand the relative importance for this.

**Comment by sablanc-msft** [comment]
Most async code usually runs on the ThreadPool (unless you use the TaskCreationOptions.LongRunning option). When you call Thread.Sleep in a async method, you are blocking a thread from the ThreadPool (which have more overhead than regular threads due to their associated local queues and other factors). Due to this overhead, the ThreadPool tries to keep the number of threads low (defaults to the number of cores) and only creates a new thread to try to keep up with the queue of tasks to process (increase throughput) every 500 ms. 

Due to this delay in the creation it is very easy to incur in ThreadPool starvation if you block ThreadPool threads. This issue builds up over time and can bring down entire production services, and I know first hand of multiple instances of this happening in production. Some developers try to increase the minimum number of threads in the ThreadPool to a large number to skip this delay, but that has other undesired side effects because each thread has it's own task local queue and tries to steal work from other threads when its local queue is empty and this creates an n2 overhead. On top of that keeping alive a large number of threads also slows down the Garbage Collector runs. All this can be easily seen using the Azure Profiler or Perfview.

Many developers are not aware of these issues and use Thread.Sleep() or any other blocking calls (like the ones issued by the CDS SDK clients, which are all synchronous) in production code. I had to fix many instances of this in our codebase when I joined my current team.
